File: golang/internal/gguf/reader.go
package gguf

import (
	"encoding/binary"
	"math"
)

type reader struct {
	bytes  []byte
	cursor int
}

func newReader(raw []byte) *reader {
	return &reader{bytes: raw}
}

func (r *reader) position() int {
	return r.cursor
}

func (r *reader) readExact(length int) ([]byte, error) {
	end := r.cursor + length
	if end < r.cursor {
		return nil, errIntegerOverflow()
	}
	if end > len(r.bytes) {
		return nil, errUnexpectedEOF()
	}
	out := r.bytes[r.cursor:end]
	r.cursor = end
	return out, nil
}

func (r *reader) readU8() (uint8, error) {
	raw, err := r.readExact(1)
	if err != nil {
		return 0, err
	}
	return raw[0], nil
}

func (r *reader) readU16() (uint16, error) { return r.readUint16() }
func (r *reader) readU32() (uint32, error) { return r.readUint32() }
func (r *reader) readU64() (uint64, error) { return r.readUint64() }

func (r *reader) readUint16() (uint16, error) {
	raw, err := r.readExact(2)
	if err != nil {
		return 0, err
	}
	return binary.LittleEndian.Uint16(raw), nil
}

func (r *reader) readUint32() (uint32, error) {
	raw, err := r.readExact(4)
	if err != nil {
		return 0, err
	}
	return binary.LittleEndian.Uint32(raw), nil
}

func (r *reader) readUint64() (uint64, error) {
	raw, err := r.readExact(8)
	if err != nil {
		return 0, err
	}
	return binary.LittleEndian.Uint64(raw), nil
}

func (r *reader) readI8() (int8, error) {
	value, err := r.readU8()
	return int8(value), err
}

func (r *reader) readI16() (int16, error) {
	value, err := r.readU16()
	return int16(value), err
}

func (r *reader) readI32() (int32, error) {
	value, err := r.readU32()
	return int32(value), err
}

func (r *reader) readI64() (int64, error) {
	value, err := r.readU64()
	return int64(value), err
}

func (r *reader) readF32() (float32, error) {
	value, err := r.readU32()
	if err != nil {
		return 0, err
	}
	return math.Float32frombits(value), nil
}

func (r *reader) readF64() (float64, error) {
	value, err := r.readU64()
	if err != nil {
		return 0, err
	}
	return math.Float64frombits(value), nil
}

func (r *reader) readString() (string, error) {
	length, err := r.readU64()
	if err != nil {
		return "", err
	}
	if length > uint64(len(r.bytes)) {
		return "", errIntegerOverflow()
	}
	raw, err := r.readExact(int(length))
	if err != nil {
		return "", err
	}
	return string(raw), nil
}

File: oxidize-python/oxidize_python/internal/gguf/reader.py
"""Binary GGUF reader mirroring oxidize-golang/internal/gguf/reader.go."""

from __future__ import annotations

import struct
from io import BufferedIOBase, BytesIO

from oxidize_python.internal.gguf.errors import (
    MAX_GGUF_STRING_BYTES,
    err_integer_overflow,
    err_string_too_long,
    err_unexpected_eof,
)


class BinaryReader:
    def __init__(self, src: BufferedIOBase | BytesIO) -> None:
        self._src = src
        self._cursor = 0

    def position(self) -> int:
        return self._cursor

    def read_exact(self, length: int) -> bytes:
        if length < 0:
            raise err_integer_overflow()
        out = self._src.read(length)
        if len(out) != length:
            raise err_unexpected_eof()
        self._cursor += length
        return out

    def read_u8(self) -> int:
        return self.read_exact(1)[0]

    def read_u16(self) -> int:
        return struct.unpack("<H", self.read_exact(2))[0]

    def read_u32(self) -> int:
        return struct.unpack("<I", self.read_exact(4))[0]

    def read_u64(self) -> int:
        return struct.unpack("<Q", self.read_exact(8))[0]

    def read_i8(self) -> int:
        return struct.unpack("<b", self.read_exact(1))[0]

    def read_i16(self) -> int:
        return struct.unpack("<h", self.read_exact(2))[0]

    def read_i32(self) -> int:
        return struct.unpack("<i", self.read_exact(4))[0]

    def read_i64(self) -> int:
        return struct.unpack("<q", self.read_exact(8))[0]

    def read_f32(self) -> float:
        return struct.unpack("<f", self.read_exact(4))[0]

    def read_f64(self) -> float:
        return struct.unpack("<d", self.read_exact(8))[0]

    def read_string(self) -> str:
        length = self.read_u64()
        if length > (1 << 63) - 1:
            raise err_integer_overflow()
        if length > MAX_GGUF_STRING_BYTES:
            raise err_string_too_long(length)
        return self.read_exact(int(length)).decode("utf-8", errors="replace")
